providers/gitlab.go

```diff
@@ -67,6 +70,12 @@ func (p *GitLabProvider) Redeem(ctx context.Context, redirectURL, code string) (
 	return
 }
 
+// SetProjects adds Projects to a GitLabProvider struct and extends OAuth2 needed scope
+func (p *GitLabProvider) SetProjects(projects []string) {
+	p.Projects = projects
+	p.Scope += " read_api"
```

Comment:
I think you need to be aware of when the enduser consciously updates the `scope` option and not clobber it when they put in custom values.

We just had to think about something similar with the OIDC provider adding group claim AuthZ:

https://github.com/oauth2-proxy/oauth2-proxy/pull/616/files#diff-2690264192c9633892c7bda108d4a58dR156

I think similar logic might be warranted here -- only add ` read_api` if `p.Scope` isn't explicitly set by the user. If I'm following this logic correctly, you are always adding this to the Gitlab scope, so should you just add `read_api` to `gitlabDefaultScope`?

Reply:
Just to clarity, this is dynamically added here (if projects options are specified) https://github.com/oauth2-proxy/oauth2-proxy/pull/630/files#diff-80edce26da4509c90a73a1669a5bbdc7R180. I made this to not break old installations without the need of this projects feature (as requested by @JoelSpeed). If something needs to be change in this behavior, tell me and i'll do it.

Reply:
I think you only add `read_api` in `SetProjects` (I agree with @JoelSpeed on his conditionality point) if the user didn't set something explicitly themselves for the `scope` 🤔 

E.g. If I explicitly set the `--scope` to `openid email read_api` this code turns that into `openid email read_api read_api`. I feel like if a user explicitly changes the default `scope` for whatever reason (even if it results in bugs for them because they forgot `read_api` after wanting this feature), at that point our code can't attempt to mutate it.

Reply:
Ok, I think I get it ! While trying to figure it out, i came across this : https://github.com/oauth2-proxy/oauth2-proxy/blob/master/pkg/validation/options.go#L153. This is what I get as default scope when creating the Gitlab Provider struct (without scope option).

I did not find a way to know if the user requested a specific scope or if it's the default one. Doing a diff is way to tricky to be maintainable.

The approach I take is : avoid duplicates `read_api` in the scope (ie: check if absent before adding it). I think it's the most viable solution.

Reply:
I still have concerns this is always called even when project authorization isn't configured: https://github.com/oauth2-proxy/oauth2-proxy/pull/630/files#diff-2690264192c9633892c7bda108d4a58dR290

The link you sent above about the conditional `SetProject` call that has this logic was in a unit test.

Reply:
Ok, i just add a check before calling SetProjects https://github.com/oauth2-proxy/oauth2-proxy/pull/630/commits/c0e9d1c1966dffc42b964f9a4f06dbe2216e4960. If everything is fine, I can squash and rebase before you merge. Thanks a lot for your feedback !